Tests/GRPCCoreTests/ServerTests.swift

```diff
@@ -0,0 +1,441 @@
+  }
+
+  func testInterceptorsAreNotAppliedToUnimplementedMethods() async throws {
+    let counter = ManagedAtomic(0)
+
+    try await self.withInProcessClientConnectedToServer(
+      services: [],
+      interceptors: [.requestCounter(counter)]
+    ) { client, _ in
+      try await client.withStream(
+        descriptor: MethodDescriptor(service: "not", method: "implemented")
+      ) { stream in
+        try await stream.outbound.write(.metadata([:]))
+        stream.outbound.finish()
+
+        let parts = try await stream.inbound.collect()
+        XCTAssertStatus(parts.first) { status, _ in
+          XCTAssertEqual(status.code, .unimplemented)
+        }
+      }
+    }
+
+    XCTAssertEqual(counter.load(ordering: .sequentiallyConsistent), 0)
+  }
+
+  func testNoNewRPCsAfterServerStopListening() async throws {
+    try await withInProcessClientConnectedToServer(services: [BinaryEcho()]) { client, server in
+      // Run an RPC so we know the server is up.
+      try await self.doEchoGet(using: client)
+
+      // New streams should fail immediately after this.
+      server.stopListening()
+
+      // RPC should fail now.
+      await XCTAssertThrowsRPCErrorAsync {
+        try await client.withStream(descriptor: BinaryEcho.Methods.get) { stream in
+          XCTFail("Stream shouldn't be opened")
+        }
+      } errorHandler: { error in
+        XCTAssertEqual(error.code, .failedPrecondition)
+      }
+    }
+  }
+
+  func testInFlightRPCsCanContinueAfterServerStopListening() async throws {
+    try await withInProcessClientConnectedToServer(services: [BinaryEcho()]) { client, server in
+      try await client.withStream(descriptor: BinaryEcho.Methods.get) { stream in
+        try await stream.outbound.write(.metadata([:]))
+
+        // New streams should fail immediately after this.
+        server.stopListening()
+
+        // Don't need to validate the response, just that the server is running.
+        try await stream.outbound.write(.message([0]))
+        stream.outbound.finish()
+
+        let parts = try await stream.inbound.collect()
+        XCTAssertEqual(parts.count, 3)
+      }
+    }
+  }
+
+  func testCancelRunningServer() async throws {
+    let inProcess = self.makeInProcessPair()
+    let task = Task {
+      var server = Server()
+      server.services.register(BinaryEcho())
+      server.transports.add(inProcess.server)
+      try await server.run()
+    }
+
+    try await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask {
+        try? await inProcess.client.connect(lazily: true)
+      }
+
+      try await self.doEchoGet(using: inProcess.client)
+      // The server must be running at this point as an RPC has completed.
+      task.cancel()
+      try await task.value
+
+      group.cancelAll()
+    }
+  }
+
+  func testTestRunServerWithNoTransport() async throws {
+    let server = Server()
+    try await server.run()
```

Comment:
Would it be reasonable to throw if we try to run a server without a transport configured?

Reply:
Good question, I thought about this too but decided against it at the time because it was easier to explain in a sentence what the function does. I think I've changed my mind (again) though: the most likely reason for not configuring a transport is a bug, we should make that more obvious so let's throw an error.